kong/tools/database_cache.lua

```diff
@@ -159,30 +165,42 @@ function _M.get_or_set(key, ttl, cb, ...)
   end
 
   -- Lock acquired. Since in the meantime another worker may have
-  -- populated the value we have to check again
+  -- populated the value we have to check again.
+  -- Errors will be collected, but only dealt with AFTER unlocking
   value = _M.get(key)
-  if value == nil then
-    -- Get from closure
-    value, err = cb(...)
-    if err then
-      return nil, err
+  if value ~= nil then
+    -- shm cache success
+    local ok, lerr = lock:unlock()
+    if not ok and lerr then
+      ngx_log(ngx.ERR, "failed to unlock: ", lerr)
     end
 
-    if value ~= nil then
-      local ok, err = _M.set(key, value, ttl)
-      if not ok then
-        ngx_log(ngx.ERR, err)
-        return
-      end
+    return value
+  end
+
+  -- cache failed, we need to invoke the callback
+  value = pack(cb(...))
+
+  if value[1] ~= nil then
+
+    local ok, err = _M.set(key, value[1], ttl)
+    if not ok then
+      ngx_log(ngx.ERR, err)
     end
   end
 
-  local ok, err = lock:unlock()
-  if not ok and err then
-    ngx_log(ngx.ERR, "failed to unlock: ", err)
+  local ok, lerr = lock:unlock()
+  if not ok and lerr then
+    ngx_log(ngx.ERR, "failed to unlock: ", lerr)
+  end
+
+  if value[1] ~= nil then
+    -- only return first value, as on each next lookup the cache will
+    -- also only return a single value
+    return value[1]
   end
 
-  return value
+  return unpack(value)
```

Comment:
> This allows the callback to return multiple return values

got that, but now if a callback does not properly return an error, we have:

```
return nil, [nil]
```

And the caller does:
```lua
local _, err = get_or_set()
if err then -- nil, because this is the [nil] one returned by the get_or_set cb

end

-- we missed our chance to return HTTP 500 because the callback errored
```